go/internal/a2a/recording_manager.go

```diff
@@ -0,0 +1,200 @@
+	return &RecordingManager{
+		client:   client,
+		dbClient: dbClient,
+		agentRef: agentRef,
+	}
+}
+
+// getUserID extracts user ID from context.
+// similar to _get_user_id in kagent-adk's request_converter.py
+func (m *RecordingManager) getUserID(ctx context.Context, contextID string) string {
+	if session, ok := auth.AuthSessionFrom(ctx); ok {
+		if session.Principal().User.ID != "" {
+			return session.Principal().User.ID
+		}
+	}
+
+	return "A2A_USER_" + contextID
+}
+
+func (m *RecordingManager) OnSendMessage(ctx context.Context, request protocol.SendMessageParams) (*protocol.MessageResult, error) {
+	if request.Message.MessageID == "" {
+		request.Message.MessageID = protocol.GenerateMessageID()
+	}
+	if request.Message.Kind == "" {
+		request.Message.Kind = protocol.KindMessage
+	}
+
+	result, err := m.client.SendMessage(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	if task, ok := result.Result.(*protocol.Task); ok {
+		logger := ctrllog.FromContext(ctx).WithName("recording-manager")
+
+		if m.dbClient != nil {
+			userID := m.getUserID(ctx, task.ContextID)
+
+			// Get agent ID from the agent ref
+			var agentID *string
+			if m.agentRef != "" {
+				agentIDStr := utils.ConvertToPythonIdentifier(m.agentRef)
+				agentID = &agentIDStr
+			}
+
+			session := &database.Session{
+				ID:      task.ContextID,
+				UserID:  userID,
+				AgentID: agentID,
+			}
+
+			// Store a session (fail silently if duplicate)
+			if err := m.dbClient.StoreSession(session); err != nil {
+				logger.Error(err, "Failed to create session", "contextID", task.ContextID, "sessionID", session.ID, "agentID", session.AgentID)
+			}
+
+			// Store Task - always store whatever the remote agent provides
+			if err := m.dbClient.StoreTask(task); err != nil {
+				logger.Error(err, "Failed to store sync task", "taskID", task.ID, "contextID", task.ContextID)
+			}
+		} else {
+			logger.Info("No database client available for storing sync task")
+		}
+	}
+
+	return result, nil
+}
+
+// TODO: Should we handle Task creation and storage here to keep track of history?
+func (m *RecordingManager) OnSendMessageStream(ctx context.Context, request protocol.SendMessageParams) (<-chan protocol.StreamingMessageEvent, error) {
+	if request.Message.MessageID == "" {
+		request.Message.MessageID = protocol.GenerateMessageID()
+	}
+	if request.Message.Kind == "" {
+		request.Message.Kind = protocol.KindMessage
+	}
+
+	upstream, err := m.client.StreamMessage(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: If DB is not available, passthrough? or should we return an error since we won't be able to store chat history information?
+	if m.dbClient == nil {
+		return upstream, nil
+	}
+
+	out := make(chan protocol.StreamingMessageEvent)
+
+	// TODO: We shouldn't _expect_ a Task to be sent during streaming.
+	// If not, this means we'd need to have an internal Task. This way we can store it in the database for future reference.
```

Comment:
Rather than waiting for a `task` to be sent, can we instead create a task on update if one doesn't already exist. Similarly we may need to create a session, which we should do at very beginning.